src/java.base/share/classes/java/net/SocketPermission.java

```diff
@@ -651,28 +652,23 @@ private boolean isUntrusted()
      * attempt to get the fully qualified domain name
      *
      */
-    void getCanonName()
-        throws UnknownHostException
-    {
-        if (cname != null || invalid || untrusted) return;
-
+     void getCanonName() throws UnknownHostException {
+         if (cnames != null || invalid || untrusted) {
+            return;
+         }
         // attempt to get the canonical name
-
         try {
-            // first get the IP addresses if we don't have them yet
-            // this is because we need the IP address to then get
-            // FQDN.
-            if (addresses == null) {
-                getIP();
-            }
-
             // we have to do this check, otherwise we might not
             // get the fully qualified domain name
             if (init_with_ip) {
-                cname = addresses[0].getHostName(false).toLowerCase();
+                    cnames = new String[]{addresses[0].getHostName(false).toLowerCase()};
             } else {
-             cname = InetAddress.getByName(addresses[0].getHostAddress()).
-                                              getHostName(false).toLowerCase();
+               cnames = new String[addresses.length];
+                int index = 0;
+                for (InetAddress inAdd : addresses) {
+                    cnames[index++] = InetAddress.getByName(inAdd.getHostAddress()).
+                            getHostName(false).toLowerCase();
+                }
             }
         } catch (UnknownHostException uhe) {
             invalid = true;
```

Comment:
The try() catch{} should be inside the for loop because each lookup failure should be handled independently. If any of them fails then set that cnames entry to null. The check for cname in the match() method needs to check for null and return false in that case. At the end of the loop then check if at least one cname exists. If there are none, then invalid can be set to true.